kpops/utils/yaml_loading.py

```diff
@@ -37,3 +37,43 @@ def substitute(input: str, substitution: Mapping[str, Any] | None = None) -> str
     if not substitution:
         return input
     return Template(input).safe_substitute(**substitution)
+
+
+def substitute_nested(input: str, max_repetitions: int = 500, **kwargs) -> str:
+    """Allow for multiple substitutions to be passed.
+
+    Will make as many passes as needed to substitute all possible placeholders.
+    A ceiling is set to avoid infinite loops.
+
+    HINT: If :param input: is a ``Mapping`` that you converted into ``str``,
+    You can pass it as a string, and as a ``Mapping`` to enable self-reference.
+
+    :Example:
+
+    >>> substitution = {
+            "a": "0",
+            "b": "${a}",
+            "c": "${b}",
+            "d": "${a}",
+        }
+    >>> input = "${a}, ${b}, ${c}, ${d}"
+    >>> print("Substituted string: " + substitute_nested(input, **substitution))
+    0, 0, 0, 0
+
+    :param input: The raw input containing $-placeholders
+    :type input: str
+    :param max_repetitions: The max number of consecutive substitutions to be carried out, defaults to 500
+    :type max_repetitions: int, optional
+    :param **kwargs: Substitutions
+    :raises Exception: Substitution was repeated {counter} times and placeholders still exist, check for loops
+    :return: Substituted input string
+    :rtype: str
+    """
+    if not kwargs:
+        return input
+    old_str, new_str = "", substitute(input, kwargs)
+    steps: set[str] = {old_str}
+    while new_str not in steps:
```

Comment:
should we raise an exception if the substituted string is detected in the set of seen substitutions? to communicate to the user that their substitution is looping and should be fixed

Reply:
ah no, this is probably not that straightforward

Reply:
perhaps you could check if the return value still contains placeholders?

Reply:
I could add a line that compares the last 2 strings after the loop is finished. If they are different, a loop must exist

Reply:
done

if it wasn't for that, I could remove the `old_str` var completely

Reply:
I think that would happen only if `new_str` is somehow found to be in the empty set `steps`. Is that possible?

Reply:
I can revert back to assigning the empty string to `old_str`

Reply:
I don't think the scenario of it being unbound is possible. We should either silence the diagnostic or init it with an empty string as before.

Reply:
I will init it like before.

Reply:
done